### PR TITLE
chore(amplify_core): move model-related types from datastore interface to amplify_core

### PR DESCRIPTION
**Pull request by ragingsquirrel3**
IMPORTANT NOTE: Do not merge this until 0.3.0-rc.2 is released. This change is backwards-compatible so it should not break that release, but it is desired to release that preview without the change in this PR.

This PR moves model-related dependencies from datastore_plugin_interface to core so that they can be used with a codegen model that imports from core instead of datastore_plugin_interface. This is to prepare for model-based helpers so that users can use codegen without a datastore dependency. Datastore interface will continue to export the same files (imports/exports from core) so that older versions of codegen files will continue to resolve as expected.

Some other loose-ish changes here:
* made some minor changes to the moved files so they pass linting as they are now subject to linting in core.
* added `AmplifyCodeGenModelException` to core to use for similar datastore exceptions in codegen
* removed lingering UUID stuff from datastore interface in favor of core stuff
* delete an orphaned, unexported json file from core

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by ragingsquirrel3** [comment]
Callout about test coverage: Flutter test coverage down with this PR bc there are some unit tests in datastore interface that should go to core. I did not move them here bc I would need to redo the codegen with the core dependency. I would rather wait until that codegen change is available as a tagged release than do it now with a local change. As a followup, those codegen models could be moved to amplify-test or core, then the unit tests for them moved to core.

https://github.com/aws-amplify/amplify-codegen/compare/tagged-release/custom-type-preview...ragingsquirrel3:tagged-release/flutter-deps-datastore-to-core?expand=1

File: packages/amplify_core/lib/src/types/models/auth_rule.dart [hunk 1]
```diff
@@ -67,7 +67,7 @@ class AuthRule {
   }
 
   Map<String, dynamic> toMap() {
-    Map<String, dynamic> map = {
+    final map = <String, dynamic>{
```
Comment (HuiSF) [inline packages/amplify_core/lib/src/types/models/auth_rule.dart:70]: nit: `<String, dynamic>` probably is unnecessary the analyzer should be able to infer the type?

File: packages/amplify_core/lib/src/types/models/auth_rule.dart [hunk 2]
```diff
@@ -90,7 +91,7 @@ class AuthRule {
         groupsField: map['groupsField'],
         provider: map['provider'],
         operations: List<ModelOperation>.from(
-            map['operations']?.map((x) => ModelOperation.values[x])));
+            map['operations']?.map((dynamic x) => ModelOperation.values[x])));
```
Comment (HuiSF) [inline packages/amplify_core/lib/src/types/models/auth_rule.dart:94]: nit: `dynamic` probably is unnecessary here as the analyzer infers `x` as `dynamic` already?

File: packages/amplify_core/lib/src/types/models/model_association.dart [hunk 3]
```diff
@@ -46,13 +46,14 @@ class ModelAssociation {
   }
 
   Map<String, dynamic> toMap() {
-    Map<String, dynamic> map = {
+    final map = <String, dynamic>{
```
Comment (HuiSF) [inline packages/amplify_core/lib/src/types/models/model_association.dart:49]: nit: ditto

File: packages/amplify_core/lib/src/types/models/model_field.dart [hunk 4]
```diff
@@ -72,7 +72,7 @@ class ModelField {
   }
 
   Map<String, dynamic> toMap() {
-    Map<String, dynamic> map = {
+    final map = <String, dynamic>{
```
Comment (HuiSF) [inline packages/amplify_core/lib/src/types/models/model_field.dart:75]: nit: ditto

File: packages/amplify_core/lib/src/types/models/model_field_type.dart [hunk 5]
```diff
@@ -26,7 +26,7 @@ class ModelFieldType {
       {this.ofModelName, this.ofCustomTypeName});
 
   Map<String, dynamic> toMap() {
-    return {
+    return <String, dynamic>{
```
Comment (HuiSF) [inline packages/amplify_core/lib/src/types/models/model_field_type.dart:29]: nit: ditto

File: packages/amplify_core/lib/src/types/models/model_schema.dart [hunk 6]
```diff
@@ -48,21 +48,22 @@ class ModelSchema {
   }
 
   Map<String, dynamic> toMap() {
-    Map<String, dynamic> map = {
+    final map = <String, dynamic>{
```
Comment (HuiSF) [inline packages/amplify_core/lib/src/types/models/model_schema.dart:51]: nit: ditto

File: packages/amplify_core/lib/src/types/models/model_schema.dart [hunk 7]
```diff
@@ -48,21 +48,22 @@ class ModelSchema {
...
+      ..removeWhere((k, dynamic v) => v == null);
   }
 
   factory ModelSchema.fromMap(Map<String, dynamic> map) {
     return ModelSchema(
       name: map['name'],
       pluralName: map['pluralName'],
       authRules: List<AuthRule>.from(
-          map['authRules']?.map((x) => AuthRule.fromMap(x))),
+          map['authRules']?.map((dynamic x) => AuthRule.fromMap(x))),
```
Comment (HuiSF) [inline packages/amplify_core/lib/src/types/models/model_schema.dart:66]: nit: ditto

File: packages/amplify_core/lib/src/types/models/auth_rule.dart (hunk 1 above)
Comment (ragingsquirrel3) [reply packages/amplify_core/lib/src/types/models/auth_rule.dart:70]: discussed offline, but these changes were needed to satisfy the analysis linter as the files are now subject to analysis in core (but not datastore, yet).

Comment (ragingsquirrel3) [reply packages/amplify_core/lib/src/types/models/auth_rule.dart:70]: correction: @HuiSF is right about these maps that they appear to infer correctly if the `<String, dynamic>` removed (not the way it was in datastore plugin). will try to change them to infer as suggested

File: packages/amplify_core/lib/src/types/models/auth_rule.dart (hunk 2 above)
Comment (ragingsquirrel3) [reply packages/amplify_core/lib/src/types/models/auth_rule.dart:94]: discussed offline, but these changes were needed to satisfy the analysis linter as the files are now subject to analysis in core (but not datastore, yet).

File: packages/amplify_core/lib/src/types/models/model_schema.dart (hunk 6 above)
Comment (ragingsquirrel3) [reply packages/amplify_core/lib/src/types/models/model_schema.dart:51]: fixed

File: packages/amplify_core/lib/src/types/models/model_schema.dart (hunk 7 above)
Comment (ragingsquirrel3) [reply packages/amplify_core/lib/src/types/models/model_schema.dart:66]: I think this one is needed.

File: packages/amplify_core/lib/src/types/models/model_field.dart (hunk 4 above)
Comment (ragingsquirrel3) [reply packages/amplify_core/lib/src/types/models/model_field.dart]: fixed

File: packages/amplify_core/lib/src/types/models/model_association.dart (hunk 3 above)
Comment (ragingsquirrel3) [reply packages/amplify_core/lib/src/types/models/model_association.dart]: fixed

File: packages/amplify_core/lib/src/types/models/auth_rule.dart (hunk 1 above)
Comment (ragingsquirrel3) [reply packages/amplify_core/lib/src/types/models/auth_rule.dart]: fixed

**Review by HuiSF — APPROVED** [review]

**Review by Jordan-Nelson — COMMENTED** [review]
I think this PR is moving more files than it needs to. Some of the classes being moved over are datastore specific and are not used in the code generated models.

Here the files that I think should stay were they are:
- subscription_event.dart (part of public API for datastore.observe())
- query_snapshot.dart (part of public API for datastore.observeQuery())
- observe_query_throttle_options.dart (part of public API for datastore.observeQuery)
- sorted_list.dart (used internally for observeQuery)

File: packages/amplify_core/lib/amplify_core.dart [hunk 8]
```diff
@@ -19,13 +19,37 @@ library amplify_core;
...
 export 'src/types/exception/amplify_exception_messages.dart';
+export 'src/types/exception/codegen_exception.dart';
 
 /// Hub
 export 'src/types/hub/hub_channel.dart';
 export 'src/types/hub/hub_event.dart';
 export 'src/types/hub/hub_event_payload.dart';
 export 'src/types/plugin/amplify_plugin_interface.dart';
 
+// ignore: directives_ordering
```
[the remark this answers is not in the document]
Comment (ragingsquirrel3) [reply packages/amplify_core/lib/amplify_core.dart]: good suggestion, updated

File: packages/amplify_core/lib/amplify_core.dart [hunk 9]
```diff
@@ -19,13 +19,37 @@ library amplify_core;
...
+export 'src/types/temporal/datetime_parse.dart';
+export 'src/types/temporal/temporal_date.dart';
+export 'src/types/temporal/temporal_datetime.dart';
+export 'src/types/temporal/temporal_time.dart';
+export 'src/types/temporal/temporal_timestamp.dart';
+
 // Util
+export 'src/util/parsers.dart';
 export 'src/util/print.dart';
 export 'src/util/uuid.dart';
```
[the remark this answers is not in the document]
Comment (ragingsquirrel3) [reply packages/amplify_core/lib/amplify_core.dart:49]: done

**Comment by ragingsquirrel3** [comment]
@Jordan-Nelson I moved the noted files back to datastore as suggested. thanks

**Review by Jordan-Nelson — APPROVED** [review]

**Pull request merged**